File: 백준/Gold/17404. RGB거리 2/RGB거리 2.cc
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
using ai3 = array<int, 3>;

const int INF = 1e9;

int main() {
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int n; cin >> n;
    vector<ai3> cost(n);
    for(int i = 0; i < n; i++) cin >> cost[i][0] >> cost[i][1] >> cost[i][2];
    int ans = INF;
    for(int c = 0; c < 3; c++){
        vector<vector<int>> dp(n, vector<int>(3, INF));
        dp[0][c] = cost[0][c];
        for(int i = 1; i < n; i++){
            dp[i][0] = min(dp[i-1][1], dp[i-1][2])+cost[i][0];
            dp[i][1] = min(dp[i-1][0], dp[i-1][2])+cost[i][1];
            dp[i][2] = min(dp[i-1][0], dp[i-1][1])+cost[i][2];
        }
        for(int i = 0; i < 3; i++) if(i != c) ans = min(ans, dp[n-1][i]);
    }
    cout << ans;
    return 0;
}

File: 백준/Gold/17404. RGB거리 2/RGB거리 2.py
import sys
input = lambda: sys.stdin.readline().rstrip()
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)

from math import inf

def main():
    n = int(input())
    cost = [[*map(int, input().split())] for _ in range(n)]
    ans = []
    for c in range(3):
        dp = [[inf]*3 for _ in range(n)]
        dp[0][c] = cost[0][c]
        for i in range(1, n):
            dp[i][0] = min(dp[i-1][1], dp[i-1][2])+cost[i][0]
            dp[i][1] = min(dp[i-1][0], dp[i-1][2])+cost[i][1]
            dp[i][2] = min(dp[i-1][0], dp[i-1][1])+cost[i][2]
        ans.append(min(dp[-1][(c+1)%3], dp[-1][(c+2)%3]))
    print(min(ans))
    return
if __name__ == "__main__":
    main()